server/connect.go

```diff
@@ -0,0 +1,149 @@
+package main
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/mattermost/mattermost-plugin-msteams/server/store/storemodels"
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/pkg/errors"
+)
+
+const (
+	earlyAdopterThreshold = 1000
+)
+
+func (p *Plugin) botSendDirectMessage(userID, message string) error {
+	channel, err := p.apiClient.Channel.GetDirect(userID, p.userID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get bot DM channel with user_id %s", userID)
+	}
+
+	return p.apiClient.Post.CreatePost(&model.Post{
+		Message:   message,
+		UserId:    p.userID,
+		ChannelId: channel.Id,
+	})
+}
+
+func (p *Plugin) MaybeSendInviteMessage(userID string) (bool, error) {
+	if p.getConfiguration().ConnectedUsersInvitePoolSize == 0 {
+		// connection invites disabled
+		return false, nil
+	}
+
+	var nWhitelisted int
+	var pendingSince time.Time
+	now := time.Now()
+
+	user, err := p.apiClient.User.Get(userID)
+	if err != nil {
+		p.API.LogWarn("Error getting user", "user_id", userID, "error", err.Error())
+		return false, err
+	}
+
+	userInWhitelist, err := p.store.IsUserPresentInWhitelist(user.Id)
+	if err != nil {
+		p.API.LogWarn("Error getting user in whitelist", "user_id", userID, "error", err.Error())
+		return false, err
+	}
+
+	if userInWhitelist {
+		// user already connected
+		return false, nil
+	}
+
+	invitedUser, err := p.store.GetInvitedUser(user.Id)
+	if err != nil {
+		p.API.LogWarn("Error getting user invite", "user_id", userID, "error", err.Error())
+		return false, err
+	}
+
+	if invitedUser != nil {
+		pendingSince = invitedUser.InvitePendingSince
+	} else {
+		moreInvitesAllowed, n, err := p.moreInvitesAllowed()
+		if err != nil {
+			p.API.LogWarn("Error checking invite pool size", "error", err.Error())
+			return false, errors.Wrapf(err, "Error checking invite pool size")
```

Comment:
nit: propose we consistently either leave the caller to log this and keep returning the error, or we choose to log and don't return an error. (Probably the former, I'm guessing?)

Reply:
Will go with logging at the API layer, and leave these to simply return errors